File: java/PermutationSequence.java
/*
 * Filename: PermutationSequence.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Mon 18 May 2015 04:07:41 PM EDT
 */

public class Solution {
	public String getPermutation(int n, int k) {
		List<Integer> list = new ArrayList<Integer>();
		char[] seq = new char[n];
		k--;

		for (int i = 1; i <= n; i++) {
			list.add(i);
		}

		for (int i = n; i > 1; i--) {
			int fact = factorial(i-1);
			int m = k / fact;
			k = k % fact;
			seq[n-i] = (char) (list.remove(m) + '0');
		}

		seq[n-1] = (char) (list.remove(0) + '0');
		return String.valueOf(seq);
	}

	private int factorial(int n) {
		int res = 1;
		for (int i = 2; i <= n; i++) {
			res *= i;
		}
		return res;
	}
}

File: cpp/permutationSequence.cpp
/*
 * Filename: permutationSequence.cpp
 * Author: Jun Xu
 * Email: junx@cs.bu.edu
 * Created Time: Tue 26 Jan 2016 10:35:39 PM EST
 * Description: 
 */

#include <iostream>

using namespace std;

class Solution
{
	public:
		string getPermutation(int n, int k)
		{
			vector<int> vec;
			for (int i = 1; i <= n; i++)
			{
				vec.push_back(i);
			}

			string res = "";
			k--;
			for (int i = n; i > 0; i--)
			{
				int fac = factorial(i-1);
				int idx = k / fac;
				k %= fac;
				res += to_string(vec[idx]);
				vec.erase(vec.begin() + idx);
			}

			return res;
		}

	private:
		int factorial(int n)
		{
			if (n == 0)
			{
				return 1;
			}

			return (factorial(n-1) * n);
		}
};
